[yikyak02] iP

yikyak02 opened the pull request:
### **Alex - Your Friendly Task Manager Bot**

> **“Your mind is for having ideas, not holding them.” – David Allen**

Alex is here to help you manage your tasks efficiently. It's:

- Text-based 📝
- Easy to use
- ~~FAST~~  SUPER FAST ⚡
- Getting Started

[Download the latest version here](https://github.com/nus-cs2103-AY2425S1/ip/pull/531#).
Double-click the file to run it.
Start adding your tasks.
Let Alex manage them for you! 😄

Features:

- [ ] Manage your tasks

- [ ]  Track deadlines and events

- [ ]  Set reminders (coming soon?)

 
 Here's the _main_ method:
```java
public static void main(String[] args) {
       new Alex("data/tasks.txt").run();
    }
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Commit:
Finished level 0. Rename, Greet, Exit

Commit:
Level 1: Echo done

Commit:
Level 2: Add list done with tell user joke function

Commit:
Level 3: Mark as done

Commit:
Level 3: Mark as done

Commit:
Level 4: ToDos, Events, Deadlines

Commit:
A-TextUiTesting: Automated Text UI Testing

Commit:
Level 5. Handle Errors

Commit:
Level 5. Handle Errors revised

Commit:
Merge remote-tracking branch 'origin/master'

# Conflicts:
#	src/main/java/Alex.java
#	src/main/java/AlexException.java

Commit:
Level 6: Delete

Commit:
Used Enums

yikyak02 committed:
A-TextUiTesting

yikyak02 committed:
A-Enums

yikyak02 committed:
Create method to ensure the data directory exists

In the existing code, tasks are stored in a file within a "data" directory. However, this directory is not automatically created if it does not already exist, leading to potential issues when trying to save tasks.

To address this, I've added a createDataDirectory method that checks if the "data" directory exists and creates it if it doesn't. This change ensures that the application will always have a valid location to save tasks, preventing runtime errors related to missing directories.

This approach is chosen to improve the reliability of file operations, particularly in cases where the application is run in environments where the directory may not be present by default.

yikyak02 committed:
Save tasks to file after every modification

Task modifications are not persisted, leading to data loss when the program exits.

Persisting tasks to a file ensures that changes are saved and can be reloaded when the program restarts, providing a consistent user experience.

Let's:
* Implement a `saveTasks` method to write the task list to a file.
* Ensure tasks are saved after marking as done or not done, adding new tasks, and deleting tasks.
* Create a data directory if it doesn't exist to avoid file path issues.

Saving tasks after each modification is crucial for maintaining data integrity and ensuring that users do not lose their progress.

yikyak02 committed:
Add method to load tasks from file at startup

Tasks were not being persisted across sessions, leading to loss of data every time the program restarted.

Loading tasks from a file ensures that users can continue from where they left off in the previous session, enhancing user experience and data consistency.

Let's:
* Implement a `loadTasks` method to read tasks from a file at startup.
* Parse the saved task data, reconstruct task objects, and add them to the task list.
* Handle different task types (Todo, Deadline, Event) during the load process, ensuring accurate restoration of task details.
* Ensure that tasks marked as done are correctly restored with their completed status.

This change provides continuity for users by allowing them to resume their task management without losing progress.

yikyak02 committed:
Implement date and time handling for chatbot

- Updated `Deadline` class to use `java.time.LocalDateTime` for storing deadline dates and times.
- Enhanced `Deadline` and `Event` classes to parse and format dates and times using `DateTimeFormatter`.
- Modified `toString()` methods in `Deadline` and `Event` classes to display dates in user-friendly formats (e.g., MMM dd yyyy).
- Updated `saveTasks()` method to save deadlines and events in a consistent date-time format (yyyy-MM-dd HH:mm).

Features:
- Dates are now stored as `LocalDateTime` or `LocalDate` objects.
- Dates and times are formatted and displayed in a more readable format.

yikyak02 committed:
Merge branch 'branch-Level-8'

* branch-Level-8:
  Implement date and time handling for chatbot

aditig0305 commented on src/main/java/Alex.java:
```diff
@@ -0,0 +1,216 @@
+import java.io.BufferedReader;
```
Should the classes be packaged?

aditig0305 commented on src/main/java/Alex.java:
```diff
@@ -0,0 +1,216 @@
+public class Alex {
+
+    private static final ArrayList<Task> tasks = new ArrayList<>();
+    private static final String FILE_PATH = "./data/tasks.txt";
```
Nice, you followed the conventions for naming a constant.

aditig0305 commented on src/main/java/Alex.java:
```diff
@@ -0,0 +1,216 @@
+                    tasks.add(newTask);
+                    printTaskAdded(newTask);
+                    saveTasks();  // Save tasks after update
+                } else if (userInput.toLowerCase().startsWith("event ")) {
+                    String[] parts = userInput.substring(6).split(" /from | /to ");
+                    Task newTask = new Event(parts[0], parts[1], parts[2]);
+                    tasks.add(newTask);
+                    printTaskAdded(newTask);
+                    saveTasks();  // Save tasks after update
+                }
+                 else if (userInput.toLowerCase().startsWith("event ")) {
```
Is the else-if for "event " repeated?

aditig0305 commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,31 @@
+public class Task {
+    protected String description;
+    protected boolean isDone;
```
Nice, the naming convention for boolean is followed

aditig0305 commented on src/main/java/Alex.java:
```diff
@@ -0,0 +1,216 @@
+import java.util.ArrayList;
+import java.util.Scanner;
+import java.time.format.DateTimeFormatter;
+
```
Maybe JavaDoc comments can be added?

aditig0305 commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,34 @@
+        this.by = parseDateTime(by);
+    }
+
+    private LocalDateTime parseDateTime(String dateTime) {
```
I like that even within the method, the code is broken up in chunks so that it is easier to read.

aditig0305 reviewed (commented):
Generally easy to read.

yikyak02 committed:
refactor: Extract closely related code into separate classes for better OOP structure

- Created `Ui` class to handle interactions with the user.
- Created `Storage` class to manage loading and saving tasks to a file.
- Created `Parser` class to handle command parsing and interpretation.
- Created `TaskList` class to encapsulate operations on the list of tasks.
- Refactored `Alex` main class to use the newly created classes.
- Implemented a `Command` interface and example `ListCommand` to handle different user commands.
- Improved code maintainability and separation of concerns by adopting a more Object-Oriented design.

yikyak02 committed:
Merge branch 'More-OOP-branch'

* More-OOP-branch:
  refactor: Extract closely related code into separate classes for better OOP structure

yikyak02 committed:
Organize classes into packages for improved structure

- Moved classes into appropriate packages to improve code organization.
- Created `duke` package and subpackages such as `duke.ui`, `duke.storage`, `duke.parser`, and `duke.task` to better organize the code.
- Updated package declarations and import statements accordingly.
- Improved maintainability and clarity of the project structure.

yikyak02 committed:
Merge branch 'Add-Packages-Branch'

* Add-Packages-Branch:
  Organize classes into packages for improved structure

yikyak02 committed:
Merge branch 'add-gradle-support'

* add-gradle-support:
  build.gradle: Prevent generating a second JAR file
  Bump gradle and lib version
  Add Gradle support

yikyak02 committed:
Add JUnit test cases for TaskList and Parser classes

- Added tests for TaskList including adding and removing Todo, Deadline, and Event tasks.
- Implemented Parser tests to verify handling of valid and invalid commands.
- Updated test cases to ensure correct functionality and exception handling.

yikyak02 committed:
Merge branch 'JUnit-Test-Branch'

* JUnit-Test-Branch:
  Add JUnit test cases for TaskList and Parser classes

lavanyagarg112 commented on src/main/java/Alex/Command/AddCommand.java:
```diff
@@ -0,0 +1,21 @@
+package Alex.Command;
+
+import Alex.Exceptions.AlexException;
+import Alex.Storage.Storage;
+import Alex.Task.Task;
+import Alex.Task.TaskList;
+import Alex.Ui.Ui;
+
+public class AddCommand extends CommandBase {
+    private Task task;
+
+    public AddCommand(Task task) {
+        this.task = task;
+    }
+
+    @Override
+    public void execute(TaskList tasks, Ui ui, Storage storage) throws AlexException {
+        tasks.addTask(task);  // This should now use the updated format
+        storage.save(tasks.getAllTasks());
+    }
+}
```
I really like the way the OOP concepts are followed!

lavanyagarg112 commented on src/main/java/Alex/Command/AddDefaultCommand.java:
```diff
@@ -0,0 +1,26 @@
+import Alex.Storage.Storage;
+import Alex.Task.DefaultTask;
+import Alex.Task.TaskList;
+import Alex.Ui.Ui;
+
+
+public class AddDefaultCommand extends CommandBase {
+    private final DefaultTask task;
+
```
Maybe you can consider adding java doc?

lavanyagarg112 commented on src/main/java/Alex/Command/MarkCommand.java:
```diff
@@ -0,0 +1,33 @@
+
+public class MarkCommand extends CommandBase {
+    private int index;
+    private boolean markAsDone;
```
I think it might be better to follow the coding standard and use something like isDone since this is a boolean?

lavanyagarg112 commented on src/main/java/Alex/Task/Deadline.java:
```diff
@@ -0,0 +1,33 @@
+import java.time.format.DateTimeFormatter;
+
+public class Deadline extends Task {
+    public LocalDateTime by;
```
maybe the variable name could be a little more descriptive?

lavanyagarg112 reviewed (commented):
Overall looks quite neat and easy to follow!

yikyak02 committed:
Add Jar File to project.

yikyak02 committed:
Merge branch 'Jar-File-Branch'

* Jar-File-Branch:
  Add Jar File to project.

yikyak02 committed:
Add Javadoc comments to all classes

- Added detailed Javadoc comments to all classes
- Improved documentation to clarify class purposes, method functionalities, and parameter details.

This commit enhances code readability and provides clear descriptions for better maintainability and understanding.

yikyak02 committed:
Tweak import statement orders to comply with coding standards
Reordered import statements within Alex.Storage package to maintain consistent alphabetical order.

yikyak02 committed:
Revert "Tweak import statement orders to comply with coding standards"

This reverts commit 286ae6f05ca14acfa0e47b1b44474ad507c8846b.

yikyak02 committed:
Add find functionality to search for tasks by keyword

- Implemented `FindCommand` class to handle searching for tasks by keyword.
- Updated `TaskList` class with `findTasks` method to filter tasks based on the keyword.
- Modified `Parser` to recognize and parse the 'find' command.
- Updated `Ui` class to display matching tasks or a no-match message.

yikyak02 committed:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Add find functionality to search for tasks by keyword

yikyak02 committed:
Merge branch 'branch-A-JavaDoc'

* branch-A-JavaDoc:
  Revert "Tweak import statement orders to comply with coding standards"
  Tweak import statement orders to comply with coding standards Reordered import statements within Alex.Storage package to maintain consistent alphabetical order.
  Add Javadoc comments to all classes

# Conflicts:
#	data/tasks.txt
#	src/main/java/Alex.java
#	src/main/java/Alex/Parser/Parser.java
#	src/main/java/Alex/Task/TaskList.java
#	src/main/java/Alex/Ui/Ui.java

yikyak02 committed:
A-CheckStyle: Integrate CheckStyle for code quality

- Added CheckStyle configuration to detect and enforce coding style violations.
- Followed SE-EDU/guides for integrating and using CheckStyle in the project.

yikyak02 committed:
Merge branch 'A-CheckStyle-branch'

* A-CheckStyle-branch:
  A-CheckStyle: Integrate CheckStyle for code quality

yikyak02 committed:
Add GUI support to Alex chatbot

- Implement JavaFX for graphical user interface
- Integrate MainWindow.fxml for user interaction
- Link Alex chatbot instance to GUI controller
- Enable user input and response display through GUI
- Update Alex to handle commands via GUI instead of console

yikyak02 committed:
Merge branch 'branch-Level-10'

* branch-Level-10:
  Add GUI support to Alex chatbot

yikyak02 committed:
fAdd varargs support to addTasks method

- Refactored addTask method to addTasks, allowing multiple tasks to be added at once using varargs.
- Updated response generation to reflect the addition of multiple tasks.
- Improved flexibility for task additions by supporting variable-length arguments.

yikyak02 committed:
Merge branch 'Varags-branch'

* Varags-branch:
  fAdd varargs support to addTasks method

yikyak02 committed:
Add greeting on startup and farewell before exiting

- Ensure Alex chatbot greets the user upon initialization before any user input
- Implement farewell message display before exiting the application with a 1-second delay

yikyak02 committed:
Add assertions to Alex and Storage classes for validation and error handling

- Added assertions in Alex.java to ensure correct initialization and task processing.
- Enhanced Storage.java with assertions to validate file paths, task descriptions, task object creation, and date handling for Deadline and Event tasks.
- Ensured the data directory is properly created and exists after creation.
- These changes improve the robustness and reliability of the task loading and saving processes.

yikyak02 committed:
Refactor Storage class to improve code quality and readability

- Refactored the `Storage` class to reduce excessive nesting by extracting methods for parsing and creating tasks.
- Added comments to the `Storage` class to explain the purpose and functionality of methods, improving code readability.
- Updated `Main` class with comments to clarify the setup and initialization of the JavaFX application.

The refactoring aims to enhance maintainability by simplifying method logic and reducing deep nesting. Comments were added to provide clear explanations of what and why the code is doing, rather than how.

yikyak02 committed:
add comments for stream feature

yikyak02 committed:
Merge pull request #3 from yikyak02/branch-A-Streams

add comments for stream feature

yikyak02 committed:
Merge branch 'master' of https://github.com/yikyak02/ip

* 'master' of https://github.com/yikyak02/ip:
  add comments for stream feature

yikyak02 committed:
Merge branch 'branch-A-CodeQuality'

* branch-A-CodeQuality:
  Refactor Storage class to improve code quality and readability

yikyak02 committed:
Merge branch 'branch-A-Assertions'

* branch-A-Assertions:
  Add assertions to Alex and Storage classes for validation and error handling

# Conflicts:
#	src/main/java/Alex/Storage/Storage.java

yikyak02 committed:
Add Fixed Duration Task Extension

- Implemented FixedDurationTask class extending Task.
- Added support for storing and managing tasks with fixed durations.
- Updated storage handling to include duration information when saving and loading tasks.
- Enhanced UI to display and handle fixed duration tasks appropriately.
- Updated tests to cover new functionality and ensure proper task management.

This update enables the chatbot to handle tasks that have a fixed duration, expanding its capability to manage a broader range of tasks.

yikyak02 committed:
Merge branch 'branch-A-addExtension'

* branch-A-addExtension:
  Add Fixed Duration Task Extension

yikyak02 committed:
Deadline class: Fix date parsing bugs and handle invalid input

The Deadline class crashes when an invalid date format (e.g., 2024-09-09 15) is provided by the user. It also fails to handle natural language dates like "Sunday" correctly.

Fixes the issue by improving the parseDateTime() method to catch invalid date formats and throw a custom AlexException.
Implements natural language handling for common days like "today", "tomorrow", and days of the week (e.g., "Sunday").

Adds user-friendly error messages to guide users when invalid inputs are detected.
These changes ensure that the program does not crash due to invalid input and provides clear feedback to users.

yikyak02 committed:
Merge branch 'branch-fixDeadlinebugs'

* branch-fixDeadlinebugs:
  Deadline class: Fix date parsing bugs and handle invalid input

yikyak02 committed:
Fix mark and unmark command output format

Correct the display format for marking tasks as done or not done.

The MarkCommand class is updated to pass the Task object to the Ui class, ensuring that the task's current status is included in the output message. The Ui class now formats the output to display the task's status correctly.

Adjust MarkCommand to pass the Task object for proper formatting
Update Ui class to format and display task details correctly
This change ensures that the output for marking and unmarking tasks reflects the actual task status and details in the desired format.

yikyak02 committed:
Add README file and product website

Include a README.md file to provide project overview, installation instructions, and usage guidelines. Also, add a basic product website to showcase the features of the application.

The README helps new users understand how to set up and use the project, while the product website serves as a landing page for potential users.

Add README.md with setup and usage instructions
Create product website with basic layout and feature highlights
These additions improve the project's accessibility and visibility for both developers and users.

yikyak02 committed:
Update README.md

yikyak02 committed:
Merge branch 'master' of https://github.com/yikyak02/ip

* 'master' of https://github.com/yikyak02/ip:
  Update README.md

yikyak02 committed:
Rename boolean variable markAsDone to isDone in MarkCommand for clarity

- Update the boolean variable 'markAsDone' to 'isDone' in MarkCommand.java
- The new name aligns with boolean naming conventions, making the code more readable.
- Verifiy that related methods and logic now reference 'isDone' consistently.

yikyak02 committed:
improve README content formatting

- Enhance formatting for better readability
- Adjust heading structures and spacing
- Reformat sections for clearer presentation

yikyak02 committed:
apply SLAP by splitting methods in Parser class

parse method is presented as a long chunk of code to handle user input

method is slightly long and verbose

- Extract parsing logic for todo, deadline, event, and other commands
  into separate methods

- Improve code readability and maintainability by adhering to the
  Single Level Abstraction Principle (SLAP).
- Each command parsing logic now exists at a single level of abstraction,
  promoting cleaner and more modular design.

This refactor enhances the structure of the Parser class, making it easier
to manage and extend in the future.

yikyak02 committed:
Touch up on readme GUI

some of the ReadME sections have spelling errors, repetitions etc.

edited and deleted unecessary parts, refined the ReadME

improve user readability and understanding of chatbot functionalities

yikyak02 committed:
Final touch up of ReadME,md before release

ReadME is slightly incomplete, not informing user how to do download the program to use the chatbot.

User needs to know where to find the jar file to be able to download it.

ReadME is more complete after making the edits

yikyak02 committed:
Remove tasks.txt from final commit before release

tasks.txt is in the github repo as it is used to test the chatbot functionalities previously

tasks.txt should be blank, and new for users who want to use Alex

tasks.txt is removed in this commit so users can start with blank task list.

yikyak02 committed:
Add  and touchup JavaDoc comments before final release

Some of the javadoc comments are missing for some methods

Javadoc comments improve understanding of how some methods work

yikyak02 committed:
Add javadoc comments for Test classes

Test classes are not annotated with javadoc comments

Javadoc comments ensure better readability and understanding when testing

Add Javadoc comments to methods in test

yikyak02 committed:
delete extra test files

some of the Junit test files were deplicated accidentally

delete the files to make sure users are not confused
